### PR TITLE
operator: Don't remove existing pools for mgr/prometheus/rbd_stats_pools

### LINKED ISSUE #11056: Allow disabling the automatic configuration of `mgr/prometheus/rbd_stats_pools`.
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Currently, the Rook operator sets the `mgr/prometheus/rbd_stats_pools` configuration to include any Ceph pool that corresponds to a `CephBlockPool` resource that has `enableRBDStatus: true`. This is generally good but causes problems if Rook's pools are not the only ones in the cluster. If no `CephBlockPool` resources exist, it simply unsets the configuration entirely.

It should be possible to disable managing this item, perhaps in the `CephCluster` resource's `monitoring` property.

**What is use case behind this feature:**
I'm using a Proxmox managed Ceph cluster which has a variety of pools. I set `mgr/prometheus/rbd_stats_pools` manually and include both the Rook managed pools _and_ Proxmox's pools. Right now I would have to define a CephBlockPool for all of Proxmox's pools as well and that's somewhat scary, I don't want two tools managing the same pools.

**Environment**:

- Kubernetes Distribution: Rancher RKE1
- Kubernetes Version: v1.21.12
- Ceph Version: 16.2.9
- Rook Version: 1.10.1

```bash
❯ sudo ceph osd lspools
4 rbd
9 device_health_metrics
10 rbd-capacity
18 clusterfs-capacity_data
19 clusterfs-capacity_metadata
25 k8s-rbd-performance
26 k8s-rbd-capacity
27 cephfs.clusterfs-performance.meta
28 cephfs.clusterfs-performance.data
29 .rgw.root
30 default.rgw.log
31 default.rgw.control
32 default.rgw.meta
```
```bash
❯ sudo ceph config dump
WHO                MASK  LEVEL     OPTION                                 VALUE                                 RO
global                   advanced  osd_pool_default_pg_autoscale_mode     on                                      
  mon                    advanced  auth_allow_insecure_global_id_reclaim  false                                   
  mgr                    advanced  mgr/influx/hostname                    influxdb.mydomain.private             * 
  mgr                    advanced  mgr/influx/password                    _                                     * 
  mgr                    advanced  mgr/influx/port                        443                                   * 
  mgr                    advanced  mgr/influx/ssl                         true                                  * 
  mgr                    advanced  mgr/influx/username                    ceph                                  * 
  mgr                    advanced  mgr/prometheus/rbd_stats_pools         k8s-rbd-capacity,k8s-rbd-performance  * 
  mgr                    advanced  mgr/prometheus/standby_behaviour       default                                 
  mgr                    advanced  mgr/zabbix/identifier                  pve1.mydomain.local                   * 
  mgr                    advanced  mgr/zabbix/zabbix_host                 zabbix.mydomain.private               * 
  mgr                    advanced  mgr/zabbix/zabbix_sender               /usr/bin/zabbix_sender                * 
    mgr.pve1             advanced  mgr/prometheus/server_addr             10.10.0.1                             * 
    mgr.pve2             advanced  mgr/prometheus/server_addr             10.10.0.2                             * 
    mgr.pve3             advanced  mgr/prometheus/server_addr             10.10.0.3                             * 
    mgr.pve4             advanced  mgr/prometheus/server_addr             10.10.0.4                             * 
    mgr.pve5             advanced  mgr/prometheus/server_addr             10.10.0.5                             * 
  osd                    dev       bluestore_fsck_quick_fix_on_mount      true                                    
```

### PR DESCRIPTION
**Pull request by avanthakkar**
Signed-off-by: Avan Thakkar <athakkar@redhat.com>

**Description of your changes:**
Rook currently knows rbd pools only if they defined through `CephBlockPool`, so it overrides the config 
`mgr/prometheus/rbd_stats_pools` which may have already been set for pools external to Rook. Check config  before running set command and if there some existing pools append them to `enableStatsForCephBlockPools`.

**Which issue is resolved by this Pull Request:**
Resolves #11056

**Checklist:**

- [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure)).
- [ ] **Skip Tests for Docs**: If this is only a documentation change, add the label `skip-ci` on the PR.
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/pool/controller.go [hunk 1]
```diff
@@ -422,23 +422,33 @@ func configureRBDStats(clusterContext *clusterd.Context, clusterInfo *cephclient
...
 	}
-	logger.Debugf("RBD per-image IO statistics will be collected for pools: %v", enableStatsForPools)
 	monStore := config.GetMonStore(clusterContext, clusterInfo)
+	// Check for existing rbd stats pools
+	existingRBDStatsPools, e := monStore.Get("mgr", "mgr/prometheus/rbd_stats_pools")
+	if e != nil {
+		return errors.Wrapf(e, "failed to get rbd_stats_pools")
+	}
+
+	enableStatsForPools := strings.Join(enableStatsForCephBlockPools, ",") + existingRBDStatsPools
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller.go:443]: The challenge with this approach is that `configureRBDStats()` is called every time there is a pool reconcile, which also happens again every time the operator is restarted, or an existing pool is edited. We can't keep appending the same pool names, so we will need a smarter merge. What about the following?
- Filter out the pools to append if they are already in the list
- When a pool is being deleted, make sure that pool is deleted from the list. `configureRBDStats()` is called in two places, when a pool is either created or deleted. When it is deleted, pass the name of the pool that is being deleted to `configureRBDStats()` (and it will be blank called called during pool creation). 

We will also need unit tests to check these scenarios.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/pool/controller.go (hunk 1 above)
Comment (avanthakkar) [reply pkg/operator/ceph/pool/controller.go:443]: > The challenge with this approach is that `configureRBDStats()` is called every time there is a pool reconcile, which also happens again every time the operator is restarted, or an existing pool is edited. We can't keep appending the same pool names, so we will need a smarter merge. What about the following?
> 
I think we need to append same pool names every time bcz on config set it overrides the existing list of pool names, so I think the only way to resolve this is to check config for existing rbd stat pool names and set config with that list of names.
>     * Filter out the pools to append if they are already in the list
> 
>     * When a pool is being deleted, make sure that pool is deleted from the list. `configureRBDStats()` is called in two places, when a pool is either created or deleted. When it is deleted, pass the name of the pool that is being deleted to `configureRBDStats()` (and it will be blank called called during pool creation).
> 
Maybe that's already taken care of?
 https://github.com/rook/rook/blob/master/pkg/operator/ceph/pool/controller.go#L247 & 
https://github.com/rook/rook/blob/master/pkg/operator/ceph/pool/controller.go#L293
> 
> We will also need unit tests to check these scenarios.

File: pkg/operator/ceph/pool/controller.go [hunk 2]
```diff
@@ -422,23 +422,44 @@ func configureRBDStats(clusterContext *clusterd.Context, clusterInfo *cephclient
 	logger.Debug("configuring RBD per-image IO statistics collection")
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller.go:422]: > Maybe that's already taken care of?
> https://github.com/rook/rook/blob/master/pkg/operator/ceph/pool/controller.go#L247 &
> https://github.com/rook/rook/blob/master/pkg/operator/ceph/pool/controller.go#L293

Those links just point to the calling of this `configureRBDStats()` method. I don't see where this method will remove the pool from the list. All pools already in the list will remain in the list. So we will need a parameter to this method during deletion that tells which pool to remove from the list.

File: pkg/operator/ceph/pool/controller.go [hunk 3]
```diff
@@ -422,23 +422,44 @@ func configureRBDStats(clusterContext *clusterd.Context, clusterInfo *cephclient
...
 	namespaceListOpt := client.InNamespace(clusterInfo.Namespace)
 	cephBlockPoolList := &cephv1.CephBlockPoolList{}
-	var enableStatsForPools []string
+	var enableStatsForCephBlockPools []string
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller.go:425]: Unit tests still needed, you could likely just add more testing to `TestConfigureRBDStats()`.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/pool/controller.go (hunk 2 above)
Comment (avanthakkar) [reply pkg/operator/ceph/pool/controller.go:422]: I think this issue should be resolved from Ceph side itself, as any user setting this config should have this functionality of avoiding duplicates and non-existing entries. I've raised a PR for the same https://github.com/ceph/ceph/pull/48398

Comment (travisn) [reply pkg/operator/ceph/pool/controller.go:422]: If we wait for a fix from Ceph, Rook will need to behave differently for versions of ceph with or without the fix. Rook supports v16.2.0 and v17.2.0. I'd propose we can fix Rook without waiting for the ceph change. 

To remove the deleted pool from the list, the caller could pass the deleted pool name as a parameter to `configureRBDStats()`. The caller in the create/update case would just pass an empty value. Then inside this method, if the deleted pool name is set it would be removed from the list.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/pool/controller_test.go [hunk 4]
```diff
@@ -618,17 +621,39 @@ func TestConfigureRBDStats(t *testing.T) {
...
+	// Case 6: Deleted CephBlockPool should be excluded from config
+	context.Client = fake.NewClientBuilder().WithScheme(s).WithRuntimeObjects(objects...).Build()
+	err = configureRBDStats(context, clusterInfo, "my-pool-with-rbd-stats")
 	assert.Nil(t, err)
 
-	// Case 5: Two CephBlockPools with EnableRBDStats:false & EnableRBDStats:true.
+	rbdStatsPools, err = monStore.Get("mgr", "mgr/prometheus/rbd_stats_pools")
+	assert.Nil(t, err)
+	assert.Equal(t, "pool1,pool2", rbdStatsPools)
+
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller_test.go:647]: Need a test that duplicates are removed

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/pool/controller_test.go (hunk 4 above)
Comment (avanthakkar) [reply pkg/operator/ceph/pool/controller_test.go:677]: Done!

File: pkg/operator/ceph/pool/controller.go [hunk 5]
```diff
@@ -418,27 +418,64 @@ func deletePool(context *clusterd.Context, clusterInfo *cephclient.ClusterInfo,
...
 	monStore := config.GetMonStore(clusterContext, clusterInfo)
+	// Check for existing rbd stats pools
+	existingRBDStatsPools, e := monStore.Get("mgr", "mgr/prometheus/rbd_stats_pools")
+	if e != nil {
+		return errors.Wrapf(e, "failed to get rbd_stats_pools")
+	}
+
+	existingRBDStatsPoolsList := strings.Split(existingRBDStatsPools, ",")
+	enableStatsForPools := append(enableStatsForCephBlockPools, existingRBDStatsPoolsList...)
+	enableStatsForPools = unique(enableStatsForPools)
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller.go:470]: nit: how about this method name?
```suggestion
	enableStatsForPools = removeDuplicates(enableStatsForPools)
```

File: pkg/operator/ceph/pool/controller_test.go [hunk 6]
```diff
@@ -618,17 +619,71 @@ func TestConfigureRBDStats(t *testing.T) {
...
+	// Case 6: Deleted CephBlockPool should be excluded from config
+	context.Executor = &exectest.MockExecutor{
+		MockExecuteCommandWithTimeout: func(timeout time.Duration, command string, args ...string) (string, error) {
+			logger.Infof("Command: %s %v", command, args)
+			if args[0] == "config" && args[2] == "mgr" && args[3] == "mgr/prometheus/rbd_stats_pools" {
+				if args[1] == "set" && args[4] != "" {
+					return "", nil
+				}
+				if args[1] == "get" {
+					return "pool1,pool2", nil
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller_test.go:662]: Instead of duplicating this whole MockExecutor, how about a variable in the mock so you can just change what it returns for each test, something like this:
```suggestion
					return mockedPools, nil
```

File: pkg/operator/ceph/pool/controller_test.go [hunk 7]
```diff
@@ -618,17 +619,71 @@ func TestConfigureRBDStats(t *testing.T) {
...
+			return "", errors.Errorf("unexpected arguments %q", args)
+		},
+	}
+	err = configureRBDStats(context, clusterInfo, "my-pool-with-rbd-stats")
+	assert.Nil(t, err)
+
+	rbdStatsPools, err = monStore.Get("mgr", "mgr/prometheus/rbd_stats_pools")
+	assert.Nil(t, err)
+	assert.Equal(t, "pool1,pool2", rbdStatsPools)
+
```
Comment (travisn) [inline pkg/operator/ceph/pool/controller_test.go:677]: 1. How about calling `configureRBDStats()` again to make sure it preserves the same values and doesn't append duplicates?
2. Where is the test for duplicates?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/pool/controller_test.go (hunk 7 above)
Comment (avanthakkar) [reply pkg/operator/ceph/pool/controller_test.go:677]: > 1. How about calling `configureRBDStats()` again to make sure it preserves the same values and doesn't append duplicates?
> 
>     2. Where is the test for duplicates?

Oops! Missed to push them;will do with the changes suggested above.

**Comment by BlaineEXE** [comment]
I believe there is still a bug in the code as implemented. If a CephBlockPool's `enableRBDStats` field is changed from `true` to `false`, the pool will still remain in the `mgr/prometheus/rbd_stats_pools` list.
